c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -202,15 +202,18 @@ impl<'tcx> TypeChecker<'tcx, '_> {
 
     /// Unify corresponding `PointerId`s in `lty1` and `lty2`.
     ///
-    /// The two inputs must have identical underlying types.  For any position where the underlying
-    /// type has a pointer, this function unifies the `PointerId`s that `lty1` and `lty2` have at
+    /// The two inputs must have compatible ([safely transmutable](are_transmutable)) underlying types.
+    /// For any position where the underlying type has a pointer,
+    /// this function unifies the `PointerId`s that `lty1` and `lty2` have at
     /// that position.  For example, given `lty1 = *mut /*l1*/ *const /*l2*/ u8` and `lty2 = *mut
     /// /*l3*/ *const /*l4*/ u8`, this function will unify `l1` with `l3` and `l2` with `l4`.
     fn do_unify(&mut self, lty1: LTy<'tcx>, lty2: LTy<'tcx>) {
-        assert_eq!(
-            self.acx.tcx().erase_regions(lty1.ty),
-            self.acx.tcx().erase_regions(lty2.ty)
-        );
+        let ty1 = lty1.ty;
+        let ty2 = lty2.ty;
+        assert!(are_transmutable(
+            self.acx.tcx().erase_regions(ty1),
+            self.acx.tcx().erase_regions(ty2),
+        ), "types not transmutable (compatible), so PointerId unification cannot be done: {ty1:?} !~ {ty2:?}");
```

Comment:
This relaxed check should go somewhere specific to `Rvalue::Cast`, not in `do_unify`, where it would affect all `Assign` and pseudo-assignment operations.  Unfortunately, the limited cast handling I implemented previously involves some hacks that will need to be backed out before this is possible.

Regarding `do_unify`, in some other comment I wrote out the subtyping rule that this whole analysis is based on: `*mut /*1*/ T <: *mut /*2*/ U` if `perms(2) ⊆ perms(1)` and `T = U`.  `do_unify` is part of the implementation of this rule.  Specifically, `do_assign(*mut /*2*/ U, *mut /*1*/ T)` asserts the constraint `*mut /*1*/ T <: *mut /*2*/ U`, `do_assign_pointer_ids(2, 1)` asserts `perms(2) ⊆ perms(1)`, and `do_unify(U, T)` asserts `T = U`.  All three functions should be left as-is so that `Assign` and pseudo-assignments still follow the proper rules.

Now, regarding casts, there are three types involved in a cast like `(q: V) = (p: T) as U`.  The relationships required between these types are `U <: V` and `T is-castable-to U`.  However, the hacky logic in the `Cast` case of `AnalysisCtxt::type_of_rvalue` does not return the actual type `U`, but instead combines parts of `T` and parts of `U` to produce a new type `X`.  As a side effect, this has to assert that `T` and `U` are sufficiently similar (prior to this branch, this meant they had to have the same pointee type).  Then `dataflow::type_check` simply checks `X <: V`, and implements no `is-castable-to` check at all (in fact it can't, since it doesn't have access to `U`).

This PR relaxes some of the assertions in `type_of_rvalue`, but leaves `type_of_rvalue` returning the mixed type `X`.  But under these relaxed checks, in some cases `X` is not similar enough to `U` to actually pass the subtyping check `X <: V`, so this branch is forced to relax some assertions in `do_unify` as well.

A cleaner way to implement this would be to have `type_of_rvalue` actually return `U`, and have `dataflow::type_check` assert exactly the required constraints `U <: V` and `T is-castable-to U`.  Specifically:
1. In `main.rs`, when populating `acx.rvalue_tys`, create `LTy`s for the target types of all `Rvalue::Cast`s.  (The TODO comment there says "certain cases", but I think you can just do all casts instead.)  This `LTy` will be the `U` type for each cast.
2. In `AnalysisCtxt::type_of_rvalue`, remove all the `Cast` cases.  Casts should instead be handled by the `self.rvalue_tys` lookup at the top (similar to array aggregates).
3. In `dataflow::type_check`, update the `Rvalue::Cast` case in `visit_rvalue` to assert the constraint that `T is-castable-to U` (looking at the `Aggregate` case, I believe `T` is `acx.type_of(op)` and `U` is the `lty` argument of `visit_rvalue`).  This should accept the same casts that are currently allowed by `type_of_rvalue`, plus the transmutability-related relaxations from this PR.
This shouldn't require any changes to `do_unify` or the rest of `Assign` handling (`U <: V` is just the normal constraint produced by the `Assign` case).

Reply:
Now that I think about it, why do we need any check here in the first place?  Why not run static analysis after typeck (`Callbacks::after_analysis`) and simply allow everything that is valid Rust?

Reply:
We also should put some of these explanations in documentation.  They're not clear at all from the code, but a lot clearer now.

Reply:
Now that #883 has been merged, 6d20f40f39d28085e2361eacf84bd90e8635dbd3 reverts this part.